core/src/poh_service.rs

```diff
@@ -2,94 +2,66 @@
-        match self {
-            PohServiceConfig::Sleep(d) => timing::duration_as_ms(d) * num_ticks,
-            _ => panic!("Unsuppported tick config"),
-        }
-    }
-}
 
 pub struct PohService {
     tick_producer: JoinHandle<()>,
 }
 
+// Number of hashes to batch together.
+// * If this number is too small, PoH hash rate will suffer.
+// * The larger this number is from 1, the speed of recording transactions will suffer due to lock
+//   contention with the PoH hashing within `tick_producer()`.
+//
+// See benches/poh.rs for some benchmarks that attempt to justify this magic number.
+pub const NUM_HASHES_PER_BATCH: u64 = 128;
+
 impl PohService {
     pub fn new(
         poh_recorder: Arc<Mutex<PohRecorder>>,
-        config: &PohServiceConfig,
+        poh_config: &Arc<PohConfig>,
         poh_exit: &Arc<AtomicBool>,
     ) -> Self {
-        // PohService is a headless producer, so when it exits it should notify the banking stage.
-        // Since channel are not used to talk between these threads an AtomicBool is used as a
-        // signal.
         let poh_exit_ = poh_exit.clone();
-        // Single thread to generate ticks
-        let config = config.clone();
+        let poh_config = poh_config.clone();
         let tick_producer = Builder::new()
             .name("solana-poh-service-tick_producer".to_string())
             .spawn(move || {
-                let poh_recorder = poh_recorder;
-                Self::tick_producer(&poh_recorder, &config, &poh_exit_);
+                if poh_config.hashes_per_tick.is_none() {
+                    Self::sleepy_tick_producer(poh_recorder, &poh_config, &poh_exit_);
+                } else {
+                    Self::tick_producer(poh_recorder, &poh_exit_);
+                }
                 poh_exit_.store(true, Ordering::Relaxed);
             })
             .unwrap();
 
         Self { tick_producer }
     }
 
-    fn tick_producer(
-        poh: &Arc<Mutex<PohRecorder>>,
-        config: &PohServiceConfig,
+    fn sleepy_tick_producer(
+        poh_recorder: Arc<Mutex<PohRecorder>>,
+        poh_config: &PohConfig,
         poh_exit: &AtomicBool,
     ) {
+        while !poh_exit.load(Ordering::Relaxed) {
+            sleep(poh_config.target_tick_duration);
+            poh_recorder.lock().unwrap().tick();
+        }
+    }
+
+    fn tick_producer(poh_recorder: Arc<Mutex<PohRecorder>>, poh_exit: &AtomicBool) {
+        let poh = poh_recorder.lock().unwrap().poh.clone();
         loop {
-            match config {
-                PohServiceConfig::Tick(num) => {
-                    for _ in 1..*num {
-                        poh.lock().unwrap().hash();
-                    }
-                }
-                PohServiceConfig::Sleep(duration) => {
-                    sleep(*duration);
+            if poh.lock().unwrap().hash(NUM_HASHES_PER_BATCH) {
+                // Lock PohRecorder only for the final hash...
+                poh_recorder.lock().unwrap().tick();
```

Comment:
This is ok b/c we release the poh lock by the time we grab the poh_recorder lock. In the future, we'll just need to be careful not to grab the poh_lock and then the poh_recorder lock, b/c the record() logic does the reverse (grabs the poh_recorder lock then the poh_lock).

Reply:
Yep, good point.  The whole Poh subsystem is a delicate dance across a field of land mines 💣